src/randomizer.py

```diff
@@ -2298,6 +2286,61 @@ def perm_inline(
     ast.ext.insert(ast.ext.index(fn), fn_def)
 
 
+# A map of all the parents this substruct is referenced from to the list of
+# references via that parent.
+ParentRefs = Dict[ca.Struct, List[ca.StructRef]]
+
+def perm_flatten_struct(
+    fn: ca.FuncDef, ast: ca.FileAST, indices: Indices, region: Region, random: Random
+) -> None:
+    """Moves fields from a nested struct into its parent struct. Only types used
+    within the given region are affected."""
+    ast_util.deepcopy_struct_defs(ast)
```

Comment:
The AST must not be modified until we're sure that no RandomizationFailure exception is thrown, or bad things will happen

Reply:
To summarize what we discussed offline, we will run into issues if the function definition AST node is touched. You suggested asserting that the function node isn't replaced.

I considered adding this assertion inside `deepcopy_struct_defs()`. I feel like this might not be necessary though. By definition, `deepcopy_struct_defs()` won't be touching the function node.

This randomization pass doesn't modify anything inside the function node until the every end, after it has ensured there won't be a RandomizationFailure. So I think we should be okay -- please let me know though if I missed something or you feel otherwise.